File: hashTable/290_word_pattern.py
class Solution:
    # O(m+n) time | O(m+n) space - where m is pattern.length and n is s.length
    def wordPattern(self, pattern: str, s: str) -> bool:
        words = s.split()
        if len(pattern) != len(words):
            return False

        charToWord, wordToChar = {}, {}
        for char, word in zip(pattern, words):
            if (char in charToWord and charToWord[char] != word) or (word in wordToChar and wordToChar[word] != char):
                return False
            charToWord[char] = word
            wordToChar[word] = char
        return True

File: hashTable/290_word_pattern.go
import "strings"

func wordPattern(pattern string, s string) bool {
    words := strings.Split(s, " ")
    if len(pattern) != len(words) {
        return false
    }

    charToWord, wordToChar := make(map[string]string), make(map[string]string)
    for i := 0; i < len(pattern); i++ {
        char, word := string(pattern[i]), words[i]
        if mapWord, ok := charToWord[char]; ok && word != mapWord {
            return false
        }
        if mapChar, ok := wordToChar[word]; ok && char != mapChar {
            return false
        }
        charToWord[char] = word
        wordToChar[word] = char
    }
    return true
}
